### ISSUE TITLE
Max-DeepLab - [thing-ID mask here must not contain stuff or crowd region]

### ISSUE THREAD

**Issue by mateodiazs** [google-research/deeplab2#49]
Hi, Thanks so much for the awesome work and thank you for the new deeplab2 repository.
I've been trying to train Max-DeepLab with my own custom dataset for a few days now and i'm a little bit stuck. 

As an initial test I've limited the number of classes to a single class **person**. 
I've then created a tf-record conversion script that builds a 2D Panoptic mask as follows. 

- Build an initial mask of some pixel value i.e 255 (same value as ignore_label ) with equal dimensions to the input image. 
- Then for every instance of person in the image I build a submask as follows `semantic_label * label_divisor + instance_id`
- I then convert this mask to `panoptic_mask.astype(np.int32).tostring()` and then follow a similar work flow as defined in the example conversion scripts that have been provided, saving this as a raw `label_format`

In dataset.py I add my DatasetDescriptor as follows (Note: the semantic label for person is `1`)

```
    num_classes=2,
    ignore_label=255,
    panoptic_label_divisor=256,
    class_has_instances_list=(1,),
```
This presumes background is a class as well as person. If i do this i am met with the following assertion. 
```
tensorflow.python.framework.errors_impl.InvalidArgumentError:  assertion failed: [thing-ID mask here must not contain stuff or crowd region.]
	 [[{{node while_1/body/_103/while_1/cond_2/then/_328/while_1/cond_2/Assert/Assert}}]]
	 [[MultiDeviceIteratorGetNextFromShard]]
	 [[RemoteCall]]
	 [[while/body/_1/IteratorGetNext]] [Op:__inference_loop_fn_127631]


```
If i make class_has_instances_list = (2,), assuming person is the second class then i am met with the following error. 

`ValueError: Dimensions must be equal, but are 3 and 2 for '{{node DeepLabFamilyLoss/MaXDeepLabLoss/einsum_3/Einsum}} = Einsum[N=2, T=DT_FLOAT, equation="bij,bkj->bik"](DeepLabFamilyLoss/MaXDeepLabLoss/GatherV2_1, DeepLabFamilyLoss/MaXDeepLabLoss/strided_slice_2)' with input shapes: [4,130,3], [4,128,2].
`

However I also tried the following. Which assumes `num_classes` means the number of panoptic classes. 

```
    num_classes=1,
    ignore_label=255,
    panoptic_label_divisor=256,
    class_has_instances_list=(1,),
```
This gives a similar error as above. 

`    ValueError: Dimensions must be equal, but are 2 and 1 for '{{node DeepLabFamilyLoss/MaXDeepLabLoss/einsum_3/Einsum}} = Einsum[N=2, T=DT_FLOAT, equation="bij,bkj->bik"](DeepLabFamilyLoss/MaXDeepLabLoss/GatherV2_1, DeepLabFamilyLoss/MaXDeepLabLoss/strided_slice_2)' with input shapes: [4,129,2], [4,128,1].

`

Finally i have gone to this assertion. And read the comment that has been placed around it. 
```
      # Filter out IDs that are not thing instances (i.e., IDs for ignore_label,
      # stuff classes or crowd). Stuff classes and crowd regions both have IDs
      # of the form panoptic_id = semantic_id * label_divisor (i.e., instance id
      # = 0)
```
I've also ensured that the instance id for every object in an image starts at at least 1. 

Any advice/help on the matter would be most welcome. 

Thanks.

**Comment by markweberdev** [comment]
Hi mateodiazs,

It's hard to do a verbal remote debugging for your case. So let me rephrase what I believe your setup should look like:

1. We use the following encoding `semantic_label * label_divisor + instance_id` throughout deeplab2. 
2. The class and instance indices are 0-based.
3. The num_classes should really correspond to the number of classes you have (not +1). Think of it as the the number of outputs you want from your softmax layer.
3A. Your ignore region is really a ignore region, and you only have a single class:
`
num_classes=1,
ignore_label=255,
panoptic_label_divisor=256,
class_has_instances_list=(0,),
`
3B. Your ignore region is a class, too. That means it has to have a class ID within the range [0, num_classes):
`
num_classes=2,
ignore_label=1,
panoptic_label_divisor=256,
class_has_instances_list=(0,),
`
4. If you have ignore regions, make sure that their semantic class (not the instance ID) corresponds to the ignore_label. 
5. Make sure all your instances start with IDs from 1.

Please let me know if that helps.

**Comment by mateodiazs** [comment]
@markweberdev 

Thanks for the help.  I managed to get it working and the results are really great. 
The problem was exactly what you had outlined in 3B. I really wanted my ignore region to be a class too but without instance. 

So i initialised my ignore_region/background to `semantic_label * label_divisor` and then continued as normal creating instance labels for people. 

Thanks once again.
